### ISSUE TITLE
RFE: Multi-Modal DDIs

### ISSUE THREAD

**Issue by poettering** [systemd/systemd#29404]
### Component

portablectl

### Describe the solution you'd like

There's interest in making "Multi-modal DDIs" a thing in systemd. This is supposed to be the combination of systemd-sysext, systemd-confext and portable service images into one DDI, that can be shipped as one and be deployed as one.

Implementation idea:
1. Extend the GPT flags field in the Discoverable Partition Spec with a few bits that indicate the "class" of a file system, with flags for sysext, for confext, for portable services (and for whole machines)
2. Make src/shared/dissect-image.c honour these flags, i.e. take the class as input, and then search only for partitions with matching flags. Update sysext/confext/pid1 and so on to specify the right class when setting up the images
3. Teach portabled to symlink images that are attached via "portablectl attach" also into /var/lib/extensions/ and /var/lib/confexts/. Then fire `systemd-sysext --refresh` to activate them.
4. Teach systemd-sysext a varlink interface so that the `systemd-sysext --refresh` can be triggered via IPC
5. To make this nicely usable on traditional distros: writable sysext stacks
6. systemd-repart should probably learn merging two DDIs for multiple classes into one multi-modal DDI.

Use-case for all of this: GNOME's sysprof would like to ship some system resources (udev rules, sysusers.d, tmpfiles.d, …) together with a portable service.

**Comment by chergert** [comment]
I assume we'd want a security boundary for loading these DDI (given file-system developers view of untrusted devices) if coming from applications but installation is basically that security boundary implicitly? We could add code to `flatpak-system-helper` to do the wiring of `systemd-sysext` bits on both install and uninstall.

For Sysprof specifically, we have a D-Bus/systemd service (`/usr/libexec/sysprofd`) which does elevated operations for us such as `perf_event_open()`, reading an unredacted `/proc/kallsyms`, tailing `turbostat` RAPL data, etc.

The Sysprof UI requests the elevated operation over D-Bus to `sysprofd` which in turn does a Polkit check for a policy known by both the UI and `sysprofd`. So we need that `.policy` merged into PolicyKit *and* that service reloaded to pick up the new policy. I think it would be okay to require that Sysprof is installed to the "System" Flatpak installation to get the `systemd-sysext` bridging otherwise you expose that into user-modifiable directories.

As for `udev` integration, I don't have needs for it in actively shipping code, but the goal would be to be able to discover certain devices from GNOME Builder (say phones, tablets, IoT, etc) and automatically setup necessary bits for `adb fastboot` or equivalents for device development. I think a number of external audio platforms require udev integration today which are used by streamers. This would certainly be easier if we had some guarantees about udev forward portability.

**Comment by bluca** [comment]
> I assume we'd want a security boundary for loading these DDI (given file-system developers view of untrusted devices) if coming from applications but installation is basically that security boundary implicitly? We could add code to `flatpak-system-helper` to do the wiring of `systemd-sysext` bits on both install and uninstall.

These images are signed and verified by the kernel (via trusted/primary/secondary/machine keyrings), so trust is established before loading them, since as you noted filesystem drivers are not safe against rogue images.

**Comment by bluca** [comment]
> 3\. Teach portabled to symlink images that are attached via "portablectl attach" also into /var/lib/extensions/ and /var/lib/confexts/. Then fire `systemd-sysext --refresh` to activate them.

I would at least add, if not make default, an alternative way to set these up: as part of the work for https://github.com/systemd/systemd/issues/24864 we have added per-unit sysext/confext, so I'd add metadata to these DDIs that specify to which service they apply. This ensures that these images only affect one (or even many, but numbered) service, rather than globally the whole system. In general it is hard to predict what dropping a file in /etc will do, but when resources are namespaced it becomes trivial: only the affected namespace. Thus the attack surface of a configuration update, and its magnitude in case things were to go wrong, are drastically reduced, and the threat model is greatly simplified.

This scheme would work great with both dbus and polkit, as both run with sandboxing options enabled, and thus already in a private mount namespace.

**Comment by poettering** [comment]
> I would at least add, if not make default, an alternative way to set these up: as part of the work for #24864 we have added per-unit sysext/confext, so I'd add metadata to these DDIs that specify to which service they apply. This ensures that these images only affect one (or even many, but numbered) service, rather than globally the whole system. In general it is hard to predict what dropping a file in /etc will do, but when resources are namespaced it becomes trivial: only the affected namespace. Thus the attack surface of a configuration update, and its magnitude in case things were to go wrong, are drastically reduced, and the threat model is greatly simplified.

This doesn't work for things like tmpfiles.d for example: you cannot sensible define a sysext for systemd-tmpfiles.service since that force the service into its own fs namespace, but that would mean it couldn't do its job, since it's supposed to create/adjust files in the host fs, not some short-lived namespace.

I think we shouldn't lose focus here. extending *other* apps is the 5th step after the 1st. Let's focus on extending the OS, which is what make @chergert's usecase work.

**Comment by bluca** [comment]
Yes, tmpfiles is inherently a host thing - and that's cool, I'm not saying we should have only one or the other. I think we should support both, and use one or the other appropriately.

**Comment by chergert** [comment]
Is it possible for a single directory tree to provide the same content as a multi-modal DDI? For example, can you get the same effect with a directory structure as a partition map?

**Comment by poettering** [comment]
nope. the when you have a dir tree we expect it to already be set up.

but of course if you want to deploy your multi-modal thing via ostree, you could ship an ostree tree with two subdirs, and then symlink one subdir into `/var/lib/portables/<myapp>` and the other into `/var/lib/extensions/<sysext>`, nothing is stopping you.

(I mean, except of course that ostree security doesn't exist, for extension that shall be merged into the OS the authentication that we do via dm-verity and stuff is kinda essential I'd claim. But that's just some philosophical disagreement over app images and whether they should be secure or not).

**Comment by poettering** [comment]
(btw, we explicitly allow symlinking ddis and dirs into the two dirs)
